File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head, ListNode* prev=nullptr){
        if(head==nullptr) return prev;
        ListNode* temp = head->next;
        head->next=prev;
        return reverseList(temp, head);
    }
    
    int pairSum(ListNode* head) {
        int sum=INT_MIN;
        ListNode* slow = head;
        ListNode* fast = head->next;
        ListNode* prev;
        while(fast->next!=nullptr){
            prev = fast;
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* half = reverseList(slow->next);
        slow->next=nullptr;
        while(half!=nullptr&&head!=nullptr){
            sum = max(sum, half->val+head->val);
            half=half->next;
            head=head->next;
        }
        return sum;
    }
};

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head, prev=None):
        if not head:
            return prev
        temp = head.next
        head.next = prev
        return self.reverseList(temp, head)
    
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow  = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
        right_half = self.reverseList(slow.next)
        slow.next = None
        
        max_sum = -float('inf')
        ptr = right_half
        while ptr and head:
            max_sum= max(max_sum, head.val+ptr.val)
            ptr = ptr.next
            head = head.next
        return max_sum
        